src/util/sighash.rs

```diff
@@ -0,0 +1,909 @@
+
+            txin.previous_output.consensus_encode(&mut writer)?;
+            script_code.consensus_encode(&mut writer)?;
+            value.consensus_encode(&mut writer)?;
+            txin.sequence.consensus_encode(&mut writer)?;
+        }
+
+        if sighash != LegacySigHashType::Single && sighash != LegacySigHashType::None {
+            self.segwit_cache().outputs.consensus_encode(&mut writer)?;
+        } else if sighash == LegacySigHashType::Single && input_index < self.tx.output.len() {
+            let mut single_enc = SigHash::engine();
+            self.tx.output[input_index].consensus_encode(&mut single_enc)?;
+            SigHash::from_engine(single_enc).consensus_encode(&mut writer)?;
+        } else {
+            zero_hash.consensus_encode(&mut writer)?;
+        }
+
+        self.tx.lock_time.consensus_encode(&mut writer)?;
+        sighash_type.as_u32().consensus_encode(&mut writer)?;
+        Ok(())
+    }
+
+    /// Compute the BIP143 sighash for any flag type.
+    pub fn segwit_signature_hash(
+        &mut self,
+        input_index: usize,
+        script_code: &Script,
+        value: u64,
+        sighash_type: LegacySigHashType,
+    ) -> Result<SigHash, Error> {
+        let mut enc = SigHash::engine();
+        self.segwit_encode_signing_data_to(
+            &mut enc,
+            input_index,
+            script_code,
+            value,
+            sighash_type,
+        )?;
+        Ok(SigHash::from_engine(enc))
+    }
+
+    /// Encode the legacy signing data for any flag type into a given object implementing a
+    /// [`std::io::Write`] trait. Internally calls [`Transaction::encode_signing_data_to`]
+    pub fn legacy_encode_signing_data_to<Write: io::Write, U: Into<u32>>(
+        &self,
+        mut writer: Write,
+        input_index: usize,
+        script_pubkey: &Script,
+        sighash_type: U,
+    ) -> Result<(), Error> {
+        if input_index >= self.tx.input.len() {
+            return Err(Error::IndexOutOfInputsBounds {
+                index: input_index,
+                inputs_size: self.tx.input.len(),
+            });
+        }
+        self.tx
+            .encode_signing_data_to(&mut writer, input_index, script_pubkey, sighash_type.into())
+            .expect("writers don't error");
+        Ok(())
+    }
+
+    /// Computes the legacy sighash for any SigHashType
+    pub fn legacy_signature_hash(
+        &self,
+        input_index: usize,
+        script_pubkey: &Script,
+        sighash_type: u32,
+    ) -> Result<SigHash, Error> {
+        let mut enc = SigHash::engine();
+        self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type)?;
+        Ok(SigHash::from_engine(enc))
+    }
+
+    #[inline]
+    fn common_cache(&mut self) -> &CommonCache {
+        Self::common_cache_minimal_borrow(&mut self.common_cache, &self.tx)
+    }
+
+    fn common_cache_minimal_borrow<'a>(
+        common_cache: &'a mut Option<CommonCache>,
+        tx: &R,
```

Comment:
Why we are passing these two arguments when we can just use `self`?

Reply:
We can't because [`self.segwit_cache` may be borrowed](https://github.com/rust-bitcoin/rust-bitcoin/pull/628/files/c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f#diff-d38c556da0f54490a5799f38921a86b42a25dc5a26df0389d7a741ce95d33e58R587) during call to this function which prevents us to borrow `self` mutably. The name of the function is supposed to indicate it but maybe it's not clear enough? Have an idea for a better one?

Reply:
Right, I see. Yes, I do not know any other workaround